File: PasswordValidation/PasswordValidation.Tests/PasswordValidatorBuilderShould.cs
using FluentAssertions;
using Xunit;

namespace PasswordValidation.Tests;

public class PasswordValidatorBuilderShould
{
    [Fact]
    public void ReturnAPasswordValidatorFromBuild()
    {
        new PasswordValidatorBuilder().Build().Should().NotBeNull();
    }
    
    [Fact]
    public void ReturnThePasswordValidatorBuilderFromAdd()
    {
        var builder = new PasswordValidatorBuilder();
        builder.Add(new TestPasswordValidator()).Should().Be(builder);
    }
    
    [Theory]
    [InlineAutoMoqData("Ab1_xxxxx", true, "password is 9 characters in length")]
    [InlineAutoMoqData("Ab1_", false, "password is < 9 characters in length")]
    [InlineAutoMoqData("ab1_xxxxx", false, "password does not contain an uppercase character")]
    [InlineAutoMoqData("AB1_XXXXX", false, "password does not contain an lowercase character")]
    [InlineAutoMoqData("Abc_xxxxx", false, "password does not contain a number character")]
    [InlineAutoMoqData("Ab1-xxxxx", false, "password does not contain an underscore character")]
    public void BuildAValidatorToCorrectlyValidateIteration1Passwords(string password, bool expectedValidity, string reason)
    {
        CompositePasswordValidator validator = new PasswordValidatorBuilder()
            .Add(new LengthPasswordValidator(9))
            .Add(new UppercasePasswordValidator())
            .Add(new LowercasePasswordValidator())
            .Add(new NumberPasswordValidator())
            .Add(new UnderscorePasswordValidator())
            .Build();
        
        validator.Validate(password).Should().Be(expectedValidity, reason);
    }
    
    [Theory]
    [InlineAutoMoqData("Ab1_xxx", true, "password is 7 characters in length")]
    [InlineAutoMoqData("Ab1_", false, "password is < 7 characters in length")]
    [InlineAutoMoqData("ab1_xxx", false, "password does not contain an uppercase character")]
    [InlineAutoMoqData("AB1_XXX", false, "password does not contain an lowercase character")]
    [InlineAutoMoqData("Abc_xxx", false, "password does not contain a number character")]
    public void BuildAValidatorToCorrectlyValidateIteration2APasswords(string password, bool expectedValidity, string reason)
    {
        CompositePasswordValidator validator = new PasswordValidatorBuilder()
            .Add(new LengthPasswordValidator(7))
            .Add(new UppercasePasswordValidator())
            .Add(new LowercasePasswordValidator())
            .Add(new NumberPasswordValidator())
            .Build();
        
        validator.Validate(password).Should().Be(expectedValidity, reason);
    }
    
    [Theory]
    [InlineAutoMoqData("Ab1_xxxxxxxxxxxxx", true, "password is 17 characters in length")]
    [InlineAutoMoqData("Ab1_", false, "password is < 17 characters in length")]
    [InlineAutoMoqData("ab1_xxxxxxxxxxxxx", false, "password does not contain an uppercase character")]
    [InlineAutoMoqData("AB1_XXXXXXXXXXXXX", false, "password does not contain an lowercase character")]
    [InlineAutoMoqData("Ab1-XXXXXXXXXXXXX", false, "password does not contain an underscore character")]
    public void BuildAValidatorToCorrectlyValidateIteration2BPasswords(string password, bool expectedValidity, string reason)
    {
        CompositePasswordValidator validator = new PasswordValidatorBuilder()
            .Add(new LengthPasswordValidator(7))
            .Add(new UppercasePasswordValidator())
            .Add(new LowercasePasswordValidator())
            .Add(new UnderscorePasswordValidator())
            .Build();
        
        validator.Validate(password).Should().Be(expectedValidity, reason);
    }
}

File: PasswordValidation/py/tests/password_validator_builder_should.py
import pytest

from password_validation.composite_password_validator import CompositePasswordValidator
from password_validation.flexible_password_validator import LengthPasswordValidator, UppercasePasswordValidator, \
    LowercasePasswordValidator, NumberPasswordValidator, UnderscorePasswordValidator
from password_validation.password_validator_builder import PasswordValidatorBuilder
from tests.always_true_password_validator import AlwaysTruePasswordValidator


def test_return_a_composite_password_validator_from_build():
    assert isinstance(PasswordValidatorBuilder().build(), CompositePasswordValidator)


def test_return_the_password_validator_builder_from_add():
    builder = PasswordValidatorBuilder()
    assert builder.add(AlwaysTruePasswordValidator()) is builder


@pytest.mark.parametrize('password, expected, reason', [
    pytest.param('Ab1_xxxxx', True, 'is >= 9 characters, has uppercase, has lowercase, has number, has underscore'),
    pytest.param('Ab1_', False, 'password is < 9 characters in length'),
    pytest.param('ab1_xxxxx', False, 'password does not contain an uppercase character'),
    pytest.param('AB1_XXXXX', False, 'password does not contain an lowercase character'),
    pytest.param('Abc_xxxxx', False, 'password does not contain a number character'),
    pytest.param('Ab1-xxxxx', False, 'password does not contain an underscore character')
])
def test_build_a_validator_to_correctly_validate_iteration1_passwords(
        password: str,
        expected: bool,
        reason: str):
    validator = PasswordValidatorBuilder() \
        .add(LengthPasswordValidator(9)) \
        .add(UppercasePasswordValidator()) \
        .add(LowercasePasswordValidator()) \
        .add(NumberPasswordValidator()) \
        .add(UnderscorePasswordValidator()) \
        .build()

    assert validator.validate(password) == expected, reason


@pytest.mark.parametrize('password, expected, reason', [
    pytest.param('Ab1_xxx', True, 'is >= 7 characters, has uppercase, has lowercase, has number'),
    pytest.param('Ab1_', False, 'password is < 7 characters in length'),
    pytest.param('ab1_xxx', False, 'password does not contain an uppercase character'),
    pytest.param('AB1_XXX', False, 'password does not contain an lowercase character'),
    pytest.param('Abc_xxx', False, 'password does not contain a number character')
])
def test_build_a_validator_to_correctly_validate_iteration2a_passwords(
        password: str,
        expected: bool,
        reason: str):
    validator = PasswordValidatorBuilder() \
        .add(LengthPasswordValidator(7)) \
        .add(UppercasePasswordValidator()) \
        .add(LowercasePasswordValidator()) \
        .add(NumberPasswordValidator()) \
        .build()

    assert validator.validate(password) == expected, reason


@pytest.mark.parametrize('password, expected, reason', [
    pytest.param('Ab1_xxxxxxxxxxxxx', True, 'is >= 17 characters, has uppercase, has lowercase, has underscore'),
    pytest.param('Ab1_', False, 'password is < 17 characters'),
    pytest.param('ab1_xxxxxxxxxxxxx', False, 'password does not contain an uppercase character'),
    pytest.param('AB1_XXXXXXXXXXXXX', False, 'password does not contain an lowercase character'),
    pytest.param('Ab1-XXXXXXXXXXXXX', False, 'password does not contain an underscore character')
])
def test_build_a_validator_to_correctly_validate_iteration2b_passwords(
        password: str,
        expected: bool,
        reason: str):
    validator = PasswordValidatorBuilder() \
        .add(LengthPasswordValidator(17)) \
        .add(UppercasePasswordValidator()) \
        .add(LowercasePasswordValidator()) \
        .add(UnderscorePasswordValidator()) \
        .build()

    assert validator.validate(password) == expected, reason
